Fix then .. catch chain

ifraixedes opened the pull request:
On the way to fix one issue that I found in this library meanwhile I was using for testing one of our implementations, I ended up having to do a quite intense refactoring in order to fix the bug, as I added new test cases which the current version doesn't pass and the first minor changes that I did to pass those new broke some of the current ones, so I did numerous changes in the implementation to pass all of them (current specs + new ones).

The bug which motivated this PR is as the one shown in the next example:

```js
SynchronousPromise.reject('something')
    .then(
            (v) => console.log(`onResolvedHandler: ${v}`),
            (e) => console.log(`onRejectedHandler: ${e}`)
    )
     .catch((e) => console.log(`catch: ${e}`))
```

The output to the console is `catch: something` and it should be `onRejectedHandler: something` as you can check using the standard `Promise` implementation.

Commit:
bug: Add test to check then onRejected handler

Add test cases which verifies if the onRejected handler is called and
when called the next catch isn't

Commit:
bug: Fix implementation of test expectation

Fix the implementation of a test expectation which was wrong according
how the promise behaves.

Commit:
bug: Fix some test specs and types on the specs descriptions

Fix some test expects implementation to check correctly what the
implementation has to do and fix some typos in the specs descriptions.

Commit:
test: Add 2 test specs more to test more cases

Add 2 specs more and their implementation to test expected behaviors in
the implementation.

Commit:
test: Use "try...catch" to always call done()

Use a "try...catch" in asynchronous expects check in order to avoid that
the `done` function isn't called if some of the expects fail.

Commit:
bug: Fix implementation to pass all the tests again.

Although this commit involved a lot of refactoring in the implementation
the main goal was to fix a bug which avoid to pass all the test specs,
because some new were added and it failed. In order to pass all of them,
not just the new ones, a implementation change to deal with the
different status of the promise and the promises chain.

ifraixedes commented:
If you want me to squash the commits in one or 2, let me know, I preferred to keep the original ones which I've done, because I believe that they help to review the changes.

fluffynuts commented:
Looks good -- I only have minor comments, which are pretty-much just around code-comments (:

I'm not a fan of code comments -- I believe that properly-refactored code shouldn't need code comments unless a person is trying to do something really strange (ie, something where the reader would exclaim "WAT?!". To this end:

1) there's a comment (twice, line 25 & line 40) about what pending means -- I'd prefer it if you could pull out a method, eg:
```
_isPendingResolutionOrRejection: function() {
  return this.status === "pending"
)
```
and remove the comment explaining what "pending" means

2) The same goes for around line 104 and line 141 -- please pull out a function, eg:
```
_findFirstResolutionHandler() {
  var next;
  while (!next && this._next.length > 0) {
    next = this._next.shift()[0];
  }
  return next;
}
```
Then the comments aren't needed and the code is cleaner (:

3) the dead code and comment at 153 can be removed

4) The comments around 168-172 are (imo) unnecessary -- they can be removed.

I'm happy with all the rest. I like the added tests and the work done to (a) make this project better and (b) keep it well-tested.

You don't have to squash all of your commits. You're also welcome to make the changes I'm asking for above with a comment about how they are to satisfy the picky project maintainer (:

Thanks for your time and effort in contributing (:

Commit:
chore: Apply code style requested by project maintainer

Apply the changes that the code maintainer (and reviewer of the last
changes) has requested according his coding style guidelines.

ifraixedes commented:
I've done the changes.

Except the _dead code_ that you have commented, which I missed to remove it before to send this PR, and for such reason I removed and amended in the commit which brought the changes in the implementation, the rest of the requested changes have been applied and committed in a new commit (the last one).

fluffynuts commented:
Looking good. Thanks for doing all the fixes (:

fluffynuts commented:
Release as 1.0.13 (:
